testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestHiveStorageFormats.java

```diff
@@ -543,6 +546,114 @@ private void runTimestampQueries(String tableName, List<TimestampAndPrecision> d
         onPresto().executeQuery("DROP TABLE " + tableName);
     }
 
+    @Test(dataProvider = "storageFormatsWithNanosecondPrecision", groups = STORAGE_FORMATS)
+    public void testStructTimestamps(StorageFormat format)
+            throws SQLException
+    {
+        setAdminRole(onPresto().getConnection());
+        ensureDummyExists();
+
+        String tableName = format("test_struct_timestamp_precision_%s_%s", format.getName().toLowerCase(Locale.ENGLISH), randomTableSuffix());
+
+        onPresto().executeQuery(format(
+                "CREATE TABLE %s ("
+                        + "   id INTEGER,"
+                        + "   arr ARRAY(TIMESTAMP),"
+                        + "   map MAP(TIMESTAMP, TIMESTAMP),"
+                        + "   row ROW(col TIMESTAMP),"
+                        + "   nested ARRAY(MAP(TIMESTAMP, ROW(col ARRAY(TIMESTAMP))))"
+                        + ") WITH (%s)",
+                tableName,
+                format.getStoragePropertiesAsSql()));
+
+        // Insert in a loop because inserting with UNION ALL sometimes makes values invisible to Presto
```

Comment:
Can you do this with `VALUES`?

Reply:
No, because Hive [doesn't allow](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DML#LanguageManualDML-Synopsis.3) functions like `array` and `named_struct` in `VALUES`.

> - Hive does not support literals for complex types (array, map, struct, union), so it is not possible to use them in INSERT INTO...VALUES clauses. This means that the user cannot insert data into a complex datatype column using the INSERT INTO...VALUES clause.